### PR TITLE
Add Gloo  SendRecv Function

### PR DESCRIPTION
**Pull request by isky-cd**
### PR types

New features 
### PR changes

OPs
### Describe

Add Gloo  SendRecv Function

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/collective/process_group_gloo.py [hunk 1]
```diff
@@ -100,6 +100,22 @@ def test_create_process_group_gloo(self):
...
+        # test send_recv
+        # rank 0
+        x = np.random.random(self.shape).astype(self.dtype)
+        tensor_x = paddle.to_tensor(x)
+        # rank 1
+        y = np.random.random(self.shape).astype(self.dtype)
+        tensor_y = paddle.to_tensor(y)
+
+        send_recv_result = paddle.assign(tensor_x)
+        if rank == 0:
```
Comment (kuizhiqing) [inline python/paddle/fluid/tests/unittests/collective/process_group_gloo.py]: * Use `pg.rank` would be better
* add send from 1 to 0

Comment (isky-cd) [reply python/paddle/fluid/tests/unittests/collective/process_group_gloo.py]: Done

File: python/paddle/fluid/tests/unittests/collective/process_group_gloo.py [hunk 2]
```diff
@@ -100,6 +101,30 @@ def test_create_process_group_gloo(self):
...
+            task = pg.send(tensor_x, 1, True)
+        else:
+            task = pg.recv(tensor_y_1, 0, True)
+            assert np.array_equal(send_recv_result_1, tensor_y_1)
+
+        if pg.rank() == 0:
+            task = pg.recv(tensor_x, 1, True)
+            assert np.array_equal(send_recv_result_2, tensor_x)
+        else:
+            task = pg.send(tensor_y_2, 0, True)
```
Comment (kuizhiqing) [inline python/paddle/fluid/tests/unittests/collective/process_group_gloo.py:125]: This part can be merged.

File: paddle/fluid/distributed/collective/send_recv.cc [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+namespace distributed {
+
+void send_recv(SendRecvOptions* opts) {
+  const auto& context = opts->context;
+  gloo::transport::UnboundBuffer* in = opts->in.get();
+  gloo::transport::UnboundBuffer* out = opts->out.get();
+  const auto slot = gloo::Slot::build(kSendRecvSlotPrefix, opts->tag);
+
+  if (context->rank == opts->src) {
+    in->send((context->rank + 1) % 2, slot);
```
Comment (sneaxiy) [inline paddle/fluid/distributed/collective/send_recv.cc]: Why `(context->rank + 1) % 2`? Same below.

Does gloo support P2P communication even if the world size of the process group is larger than 2?

Comment (isky-cd) [reply paddle/fluid/distributed/collective/send_recv.cc]: Done

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Pull request merged**